### DISCUSSION TITLE
JIT

### DISCUSSION THREAD

**Discussion by Silenus-the-Companion** [fluffos/fluffos#1107] [Ideas]
I have been thinking this over and I think it is quite a lot of work to get something that works well. The idea is to rewrite package/ops/ops.cc so that the operators don't work of the argument stack but in registers. This might include adding like with dgd float and int as "bare types". The key to this is being able to unbox ob1->f() + ob2->f() and this might require some kind of caching scheme like with the java hotspot JIT where speculatively based on the type of parameters and fields related to a method one creates an optimized version tied to the return type of ob1->f() and deoptimized to the mixed version if the object "types" don't match.

This would probably given the complexity have to be implemented in stages.

Support code -> add some sort of numerical id to program_t on compile to make it easy to identify a program_t of an object. Recycle old id's if programs share the same return types + names of methods.

Add hotspot tracking (hot methods) to the interpreter(interpret.cc/functional.cc) when functions/methods get called that bumps up a counter and has a hook for when it hits a certain. threshold to "compile" the interpreted function in the cache. This counter is specialized on the object types parameters and relevant fields of the program for the method. This second part I am not quite sure how to do. yet since not all fields are relevant since only certain fields are accessed by the method. 

Add cache/optimization/deoptimization support to this "hook". 

Add the compiler support.

**Comment by carterchengwn-byte** [comment]
I think the key to this might be to use caching of optimized code where the types are known after a guard is checked and then compile the function as a specialization where depending on the types of the parameters and relevant fields in the object the code can be made specific to those types instead of being generic with type tag checks to know what function to apply to the relevant lpc variables. the idea might be at a given call_site of an ob->f() type call to cache a series of parameter "types" (object might need to have special handling i.e. it might need to check if ob a. a->f1() a->f2() is sort of a local interface for that object. and it might need an escape hatch for cases where you have ob a passed in as a parameter a->f1() returns an object b b->f() is called (who knows what the type is?) this kind of type specialization would probably yield a nice boost in performance.
